A chplvis primer

tomyhoi opened the pull request:
- A new directory under /test/release/examples/primer called chplvis
- adds two simple test programs
- adds a chplvis.html file and associated graphic elements to provide an introduction to chplvis

lydia-duncan commented on test/release/examples/primers/chplvis/prog2.chpl:
```diff
@@ -0,0 +1,44 @@
+// Example 2 of use of VisualDebug module and chplvis tool.
```
You may want to take a look at the other primers in the directory.  They're a little more verbose and comment rich (though I think this file does get the point across).  I think it's fair to have this code live here, but the html should probably live elsewhere, like the website.

lydia-duncan replied:
I haven't looked at the html file yet (I'm doing that next).

tomyhoi replied:
Part of my idea here was to have the verbose part in the html since it is very hard to show graphics in code.   That let the code be shorter in the html.   It was also my intent that appears to have not been communicated well, that the "users" would read the html in one window and compile/execute and run chplvis in another window.     If the html ends up as part of the web site, then I'd think there shouldn't be any programs for chplvis in the /examples/primers directory, possibly just a pointer to the web site.   (The document would link to the example programs and the user would get those programs from the web site.)  If there are any files in /examples/primers, I'd expect all of them to be there.   That is why I did a subdirectory for chplvis.   I'm speaking partly from my point of view of recently needing to read the primers to refresh my memory about Chapel and to learn about new features I hadn't used before or weren't there last time I used Chapel.

lydia-duncan replied:
I think a user encountering the chplvis directory in the primers directory will first look at the .chpl files, given that is where the entirety of the other information lived.  Placing this tutorial in a different place would give you more freedom to direct the order in which the files will be read.  I think it does make sense to have the majority of the information in the .html file, especially for the call outs to the pictures provided, and I think for that reason it would be better to put this tutorial in a different place, so that its flow is properly respected.

lydia-duncan commented on test/release/examples/primers/chplvis/prog1.chpl:
```diff
@@ -0,0 +1,10 @@
+//  Example 1 using visual debug
```
This file definitely should have more comments.  Being numerically first, it is probably what will be looked at first.

tomyhoi replied:
Yes, it should point the the html file, wherever it ends up.

tomyhoi replied:
Since this file will most likely be looked at first, I added comments to point to the .html file and to explain what the code was doing.

lydia-duncan commented on test/release/examples/primers/chplvis/chplvis.html:
```diff
@@ -0,0 +1,312 @@
+compile and run the example programs and examine
+the <em>VisualDebug</em> results with <em>chplvis</em>.  The graphics
+in this primer were produced on a system using the <em>fifo</em> package
+instead of <em>qthreads</em> for the tasking layer.   If you use
```
I would suggest changing "instead of qthreads" to "instead of the default, qthreads,"

lydia-duncan commented on test/release/examples/primers/chplvis/chplvis.html:
```diff
@@ -0,0 +1,312 @@
+<!DOCTYPE HTML>
```
At a high level, I think moving this file on to the chapel website will change a lot of its formatting, so that it matches the rest of the content present there.  Someone who is more expert on that area should comment as to how thing will need to change

lydia-duncan commented on test/release/examples/primers/chplvis/chplvis.html:
```diff
@@ -0,0 +1,312 @@
+
+<h4>Example 1</h4>
+<p>
+Consider the chapel program: (prog1.chpl)
```
Hm. I was expecting a link here to the program in the release directory, but it looks like a plain copy+paste.  You may want to investigate linking it into the html file (or whatever the final file ends up being)

lydia-duncan commented on test/release/examples/primers/chplvis/chplvis.html:
```diff
@@ -0,0 +1,312 @@
+</p>
+<h4>chplvis Elements</h4>
+<ul>
+<li><em>Locale</em> is represented by a box.  The initial display
```
The English teacher in me would like an article before Locale and Data menu.  Communication links, being plural, is fine :)

lydia-duncan replied:
Or maybe "Each Locale" . . .

lydia-duncan commented on test/release/examples/primers/chplvis/chplvis.html:
```diff
@@ -0,0 +1,312 @@
+<p>
+It is important to notice the direction of the "arrow" in the header
+for the windows.   This is for communication from locale 1 to locale 0.
+The total communication was 12.  It is further broken out into three
```
I'd like this to say "The total number of communication calls was 12", or something along those lines

lydia-duncan commented on test/release/examples/primers/chplvis/chplvis.html:
```diff
@@ -0,0 +1,312 @@
+In many programs, one will want to look at a number of small parts of
+their program in addition to seeing the total statistics.  prog2.chpl
+gives an example of using the <em>VisualDebug</em> functions
+<em>tagVdebug(name)</em> and <em>pauseVdebug()</em>.
```
The italics in this section make it a little difficult for me to read

lydia-duncan commented on test/release/examples/primers/chplvis/chplvis.html:
```diff
@@ -0,0 +1,312 @@
+gives an example of using the <em>VisualDebug</em> functions
+<em>tagVdebug(name)</em> and <em>pauseVdebug()</em>.
+<pre style="margin:20px">
+// Example 2 of use of VisualDebug module and chplvis tool.
```
Same comment here about linking

lydia-duncan commented on test/release/examples/primers/chplvis/chplvis.html:
```diff
@@ -0,0 +1,312 @@
+}
+</pre>
+<p>
+Note that the <em>startVdebug("E2");</em> is places after the declarations
```
Typo on "places"?  I think it should be placed.  I would also expand "the startVdebug();" to "the startVdebug() call".

lydia-duncan commented:
I think I'm done with my observations, but I think others should take a look at it.  Overall, I'm excited to have this and pleased with what it covers :+1:

ian-bertolacci commented:
I took a look at it today.
Looks pretty good! I was able to follow along and then made my own trivial example.

For organization, I would put the tutorial and supporting files under its own subdirectory within primers/chplvis (maybe under primers/chplvis/tutorial?), that way there isnt a mess of files. But thats my own opinion. 

Eventually it would be neat to see a more complex example, or have a "Debugging Multi-Locale Programs" that used chplvis as a tool (as opposed to chplvis being the focus) to explore a more complex application. 

But I think this suffices for now. 
Looks good!

lydia-duncan commented on test/release/examples/primers/chplvis/prog1.chpl:
```diff
@@ -0,0 +1,32 @@
+
+//  Example 1 using visual debug
+//  Full primer is found in doc/chplvis/chplvis.rst.
+//  Please read that documentation for better understand chplvis.
```
"for better understand" -> "to better understand" or "for a better understanding"

lydia-duncan commented on doc/chplvis/chplvis.rst:
```diff
@@ -0,0 +1,315 @@
+same.  When the *startVdebug(name)* is executed, a collection
+of files, one per locale, are created in a directory with the *name* given
+in *startVdebug(name)*.
+</p>
```
This looks leftover from the html page

lydia-duncan commented on doc/chplvis/chplvis.rst:
```diff
@@ -0,0 +1,315 @@
+----------------
+
+- The *information* box shows the file set opened, which tag (see
+  Example 2) is displayed, and provides a color bar to help visually
```
I believe you can generate a link to the Example 2 section.

lydia-duncan commented on doc/chplvis/chplvis.rst:
```diff
@@ -0,0 +1,315 @@
+  can select number of tasks, CPU time at the locale, clock time or
+  concurrency at the locale.  Clock time is normally very close to
+  equal across all locales.  For the communication links, one can
+  select number of communications or size of data sent.</li>
```
Also looks like leftover html

lydia-duncan commented on doc/chplvis/chplvis.rst:
```diff
@@ -0,0 +1,315 @@
+gives an example of using the *VisualDebug* functions
+*tagVdebug(name)* and *pauseVdebug()*.
+
+::
```
If you're interested, the format for using a code file is

.. literalinclude:: /relative/path/to/file.chpl
   :language: chapel
   :lines: m-n (if you want to only use certain lines)

tomyhoi replied:
This doesn't work with the standard rst2html.py tool.   I'm not sure which tool does this.

lydia-duncan replied:
Sphinx, which is our common rst generator, and I believe what github uses.  If you make the change and view it through github, it ought to turn out right (but I haven't tested that myself).

tomyhoi replied:
The other issue here is the changing paths, ../../test/release/examples/primers/chplvis and just ../../examples/primers/chplvis in the release.    Would we generate a .html as part of building a release tree so the include path would be the git tree?

lydia-duncan replied:
Oh, that's a good point.  In most of our documentation, I think we use the release path and figure developers will understand what it means, but that won't work for generating the .html file.  It's probably fine to just leave it, then, though keeping the files in sync will be a pain.

lydia-duncan commented:
Looks good otherwise!

tomyhoi commented:
@lydia-duncan -- I think I've made the corrections you suggested, except the "literalinclude".   I've updated Example 3 due to a race condition I found and fixed.  I have also added another example and added another bullet to the final comments.

lydia-duncan commented on doc/chplvis/chplvis.rst:
```diff
@@ -0,0 +1,540 @@
+diverge and thus extra computation steps will not produce a "wrong"
+answer.  By doing extra computation, the result will be a bit more
+accurate.  The reader should use the *compLoop* and the *dirname*
+config variables to run several versions of this program yeilding
```
'yielding' is misspelled here

lydia-duncan commented on test/release/examples/primers/chplvis/prog3.chpl:
```diff
@@ -0,0 +1,102 @@
+  Temp(i,j) = 0.0;
+}
+
+tagVdebug("boundry");
```
Shouldn't this be "boundary"?

lydia-duncan commented on doc/chplvis/chplvis.rst:
```diff
@@ -0,0 +1,540 @@
+*tagVdebug()* calls in the program.  Selecting the tags menu
+gives the following display:
+
+.. image:: E2-2.png
```
Does this image need to be updated?  The numbering for the tags seems different than a later image.

lydia-duncan commented on doc/chplvis/chplvis.rst:
```diff
@@ -0,0 +1,540 @@
+understand the usefulness of *chplvis*.  For example, one can compare
+the CPU time used between the *computation* and *max* phases of this
+Jacobi computation.  The config variable *compLoop* allows one to run
+the computation loop more than once before than checking for convergence
```
"before than checking" -> "before then checking"?

lydia-duncan commented on doc/chplvis/chplvis.rst:
```diff
@@ -0,0 +1,540 @@
+the CPU time used between the *computation* and *max* phases of this
+Jacobi computation.  The config variable *compLoop* allows one to run
+the computation loop more than once before than checking for convergence
+in the *max* tagged code.  It is know that the Jacobi code will not
```
"It is known"?

lydia-duncan commented:
I like the new examples!  One more thing: perhaps we should add a README to the examples/primers/chplvis directory to make sure that any modifications to the test programs get propagated to the .rst file?

tomyhoi commented:
OK, README added.  Thanks for all your comments Lydia.

lydia-duncan commented:
Looks great!  Thanks!
